services/src/main/java/io/grpc/services/TempFileSink.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.services;
+
+import com.google.protobuf.MessageLite;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The output file goes to the JVM's temp dir with a prefix of BINARY_INFO. The proto messages
+ * are written serially using {@link MessageLite#writeDelimitedTo(OutputStream)}.
+ */
+class TempFileSink implements BinaryLogProviderImpl.BinaryLogSink {
+  private static final Logger log = Logger.getLogger(TempFileSink.class.getName());
+
+  private final String outPath;
+  private final OutputStream out;
+  private boolean closed;
+
+  TempFileSink() {
+    String outPath = null;
+    OutputStream out = null;
+    try {
+      File outFile = File.createTempFile("BINARY_INFO.", "");
+      outPath = outFile.getPath();
+      log.log(Level.INFO, "Writing binary logs to to {0}", outFile.getAbsolutePath());
+      out = new BufferedOutputStream(new FileOutputStream(outFile));
+    } catch (IOException e) {
+      log.log(Level.SEVERE, "caught exception creating file", e);
+    }
+    this.outPath = outPath;
+    this.out = out;
+    if (this.out == null) {
+      closed = true;
+    }
+  }
+
+  String getPath() {
+    return this.outPath;
+  }
+
+  @Override
+  public synchronized void write(MessageLite message) {
+    if (closed) {
+      return;
+    }
+    try {
+      message.writeDelimitedTo(out);
```

Comment:
This scares me. Doing any sort of blocking has typically resulted in "bad things" of various sorts. It could maybe be acceptable here because there are "flow control" concerns, but we should discuss it.